Drop statsd client from veneur's internal metrics reporting

asf-stripe opened the pull request:
<!-- Checklist For PRs

* Ensure you've written tests where applicable
* Add a CHANGELOG.md entry describing your change, thank yourself!
* Document any changes to metrics or configuration
-->


#### Summary
This PR reworks veneur such that its metrics get reported not through statsd, spewing packets at its own UDP listening port, but sending them through its own `*trace.Client` (typically internal), by attaching them to an SSF span.

#### Motivation
I'd like us to cause as little network traffic as possible, reduce syscalls, gain more clarity around what gets reported how.


#### Test plan
I made sure the tests still work, but then most of them don't exercise the path where we actually send metrics through the trace backend.

To try it out, roll this to our QA and see:
- [x] do metrics still get reported in the first place? (They should be!)
- [x] how are dashboards affected on the hosts that this is running on?
- [x] how is CPU / memory usage affected by this, possibly syscall volume too?

#### Rollout/monitoring/revert plan

This will likely affect dashboards - if you roll this out, you will notice a severe drop in UDP packets.

There is an ordering requirement for rolling it out, too:

1. Roll out local veneur servers
2. Roll out global veneur servers
3. Roll out veneur proxies

Proxies must come last because they report SSF metrics to a local veneur server, and that server must know to expect them/report them correctly - so the versions must match up

asf-stripe commented:
This is ~readyish to be looked at, but still needs some auditing for the metric names (I am sure I messed up some)!

@aditya-stripe, @cory-stripe - wanna poke? (:

cory-stripe commented on proxy.go:
```diff
@@ -282,8 +278,10 @@ func (p *Proxy) Handler() http.Handler {
 }
 
 func (p *Proxy) ProxyTraces(ctx context.Context, traces []DatadogTraceSpan) {
+	samples := []*ssf.SSFSample{}
```
Is this being used anywhere?

cory-stripe commented on server.go:
```diff
@@ -584,31 +578,32 @@ func (s *Server) HandleMetricPacket(packet []byte) error {
 func (s *Server) HandleTracePacket(packet []byte) {
-	// TODO: don't use statsd for this in favor of an in-memory counter
-	s.Statsd.Incr("ssf.received_total", nil, .1)
+	samples := []*ssf.SSFSample{ssf.Count("veneur.ssf.received_total", 1, nil)}
```
This creates a different concern. Multiple `Counts` of the same metric should likely be aggregated, no?

asf-stripe replied:
That's a good question: Should we do this in the client too or is it enough to do it in the server? Since they're relatively small, I think we could just let the server deal & introduce one fewer place for things to go wrong...

asf-stripe commented on flusher.go:
```diff
@@ -22,14 +24,14 @@ func (s *Server) Flush(ctx context.Context) {
-	s.Statsd.Gauge("mem.heap_alloc_bytes", float64(mem.HeapAlloc), nil, 1.0)
-	s.Statsd.Gauge("gc.number", float64(mem.NumGC), nil, 1.0)
-	s.Statsd.Gauge("gc.pause_total_ns", float64(mem.PauseTotalNs), nil, 1.0)
+	span.Add(ssf.Gauge("veneur.mem.heap_alloc_bytes", float32(mem.HeapAlloc), nil),
```
Because we use `float32` [in the SSF data types](https://github.com/stripe/veneur/blob/master/ssf/sample.proto#L26): it's not great. Wonder if we can bump the size of those up?

asf-stripe commented on flusher.go:
```diff
@@ -139,7 +144,7 @@ func (s *Server) tallyMetrics(percentiles []float64) ([]WorkerMetrics, metricsSu
 	}
 
-	s.Statsd.TimeInMilliseconds("flush.total_duration_ns", float64(time.Since(gatherStart).Nanoseconds()), []string{"part:gather"}, 1.0)
+	ssfmetrics.ReportOne(s.TraceClient, ssf.Timing("veneur.flush.total_duration_ns", time.Since(gatherStart), time.Nanosecond, map[string]string{"part": "gather"}))
```
oooooh, I love the idea of putting the current span on the context, though it also sounds a bit magical. Wonder what @aditya-stripe thinks about that! (:

asf-stripe commented on trace/ssfmetrics/client.go:
```diff
@@ -0,0 +1,28 @@
+// Package ssfmetrics provides routines for conveniently reporting
+// metrics attached to SSF spans.
+package ssfmetrics
```
yeah, I thought that would work, but there were some variables/functions that `metrics` would clash with, so I didn't want to conflict with those. Maybe we can free up that name though.

asf-stripe commented:
OK, I think this is ready for another look - I renamed the `ssfmetrics` package to `metrics` (and incidentally, renamed the `metrics` span sink package to `ssfmetrics`, because that is exactly that sink's job description). Also redid the way we prefix metrics, so the names should all be the same as before again!

PTAL, @cory-stripe?

asf-stripe commented:
I brought this up to date with the latest master; @cory-stripe - PTAL?

ChimeraCoder commented on ssf/samples.go:
```diff
@@ -1,6 +1,15 @@
 
-import "time"
+import (
+	"math/rand"
```
`crypto/rand` :)

asf-stripe replied:
I knew you were going to call that out, and the only defense/justification I have is that the dogstatsd library uses math/rand also, for a `0.0..1.0` RNG that mustn't fail.

ChimeraCoder replied:
`crypto/rand` is also guaranteed not to fail or block, and we generally want to be using that unless there's a compelling reason not to. Is there a reason we need deterministic behavior here?

asf-stripe replied:
soooo `crypto/rand.Int` does return an error, are you saying it's fine to ignore that? If so, I'm sure I can hack a thing together (:

asf-stripe replied:
Ah! `Sampled` was meant to be called by code constructing the samples, not by `SendSync` or any of the other sending functions - it's a constructor (that removes data, hah), rather than anything else - this is because the ideal use case for those `SSFSample`s is that you'd attach the samples you wish to send to your span directly; once you send the span, it's already too late to toss out samples.

Since this operates as an in-memory calculation, is meant to be composable, and has no expectation of even a chance failure, I think this is a place where determinism would be right.

asf-stripe replied:
I've made a benchmark for sampling a million metrics at `0.2`; looks fairly ok:

```
pkg: github.com/stripe/veneur/ssf
BenchmarkSampled-4      2000000000               0.03 ns/op
```

ChimeraCoder commented on ssf/samples.go:
```diff
@@ -47,13 +69,45 @@ func TimeUnit(resolution time.Duration) SampleOption {
+	for _, s := range samples {
+		if rand.Float32() <= rate {
+			if rate > 0 && rate <= 1 {
+				s.SampleRate = s.SampleRate * rate
```
I'm not sure I understand this part - why do we want to update the internal sample rate with each metric?

asf-stripe replied:
That's the sample rate reported to the upstream server - it needs to know by what number to multiply the measurement-as-reported in order to have an accurate estimate of the real number. Flow is as follows:

1. client calls `Sampled` on a measurement (a counter with `Value=30` `SampleRate=1`), `rate=0.5`
2. There's a 50% chance that each measurement gets dropped
3. The measurements that don't get dropped have their `SampleRate` adjusted to represent the chance that they got reported; in this example, that's `1 * 0.5`
4. The not-dropped measurements get reported to veneur, which does the inverse computation, [`Value * (1/SampleRate)`](https://sourcegraph.com/github.com/stripe/veneur@4b90543807ff43e1e8ccaecbfa4c168726d82857/-/blob/samplers/samplers.go#L133) to arrive at an estimate of a realistic counter value, taking into account the dropped measurements.

Summing up, I think this way of computing the Sample's SampleRate should allow for things like single and multiple calls of `Sampled` - e.g. `Sampled(0.2, Sampled(0.5, []*ssf.*SSFSample{ssf.Count("foo", 20, nil)}))`

cory-stripe replied:
Thanks for asking this, @aditya-stripe. I think I also wanted to but got distracted. The explanation makes sense. In effect this is the same as the "can I use different sample rates when calling `Increment`" that we get sometimes. Sure, you can since each metric is multiplied by it's sample rate if it "wins" and makes it to the Veneur's internals. Is that a correct explanation?

asf-stripe replied:
Exactly! With `Sampled`, you pre-filter the measurements, then just report whatever comes out. I haven't gotten a whole lot of examples of this in (we report at sample rate = 1 almost everywhere except the places where it could lead to amplification), but something like this is totally plausible:

``` go
span.Samples = append(span.Samples, ssf.Sampled(0.9, ssf.Count("rare.counter", 1, nil)...)...)
span.Samples = append(span.Samples, ssf.Sampled(0.01, ssf.Count("expenso.counter", 1, nil)...)...)
```

ChimeraCoder commented on flusher.go:
```diff
@@ -337,17 +345,16 @@ func (s *Server) flushForward(ctx context.Context, wms []WorkerMetrics) {
 	// the error has already been logged (if there was one), so we only care
 	// about the success case
-	if http.PostHelper(span.Attach(ctx), s.HTTPClient, s.Statsd, s.TraceClient, fmt.Sprintf("%s/import", s.ForwardAddr), jsonMetrics, "forward", true, log) == nil {
+	if http.PostHelper(span.Attach(ctx), s.HTTPClient, s.TraceClient, fmt.Sprintf("%s/import", s.ForwardAddr), jsonMetrics, "forward", true, log) == nil {
```
:D :D :D

asf-stripe commented:
OK, I think this should address all the concerns above (except the `math/rand` one which I think there's a rationale for!); PTAL everyone (:

cory-stripe commented:
👍 

Considering the scope I wouldn't mind another pass from @aditya-stripe just to get some more eyes. But it looks good to me!

asf-stripe commented:
Since this is such a beast to review, and we've gotten requests for the stats reporting functions, I'll take this back and make a new PR for those functions.

ChimeraCoder commented:
SGTM - let me know when that's ready for review! :)

asf-stripe commented:
OK, I updated this to be based on latest master and this no longer has any foundational work on it anymore, only statsd-dropping functionality!

PTAL @cory-stripe and @aditya-stripe  (-:

cory-stripe commented:
👍 

Still a go from me!

asf-stripe commented:
Update: [I rolled this to qa-obscronboxes], and - results!:

* Veneur continued to report the metrics we use for the dashboard
* There was no observable UDP traffic from veneur (I tcpdumped with and without the change)

I take that to mean that the internal metrics reporting is working!

I did not observe a noticeable drop in incoming UDP packets, as obscronboxes also run mproxy, which reports a bunch of packets/min also, but that's ok!

asf-stripe commented:
Wait, that above comment is completely invalid: I'd looked at the prod dashboard. Whoops.

There's some metrics that aren't quite right, and some that did move/change reasonably, quite a bit:

* Span received rate! This is expected!
* the metrics rate flushed by type graph has ~stabilized, it looks like? I think this might be related to us sending metrics in batches at the flush interval now.
* GC count&time went up. Guess we do allocate more data now /:

Some mistakes:
* Not tagging the signalfx sink on the flush metrics
* Lightstep sink span flush metrics don't get reported now?

I'll poke at those now.

asf-stripe commented:
Further update: I think the last commit should eliminate the last bit of write amplification! Let me tell you the saga of this bug!

In the internal client, we report the number of traces sent (so that we have a count of metrics sent internally), sampled at 10 (then, experimentally, 1) percent. Since that means the majority of metrics would not be reported, I thought that's fine. NOT FINE. The metrics client still sent an empty trace span if there were no metrics to send, again over the internal client, which again reported sampled metrics, etc. This meant that at the first chance it got, the internal client would spam one span per span (of course accounting for the capacity of the client!), which brought CPU usage to 100%. Welp!! (:

Guess this is fixed now, GC times and counts are *way* down too. So I'll take down that pooling change & keep it on a branch for when it turns out we need it. The only thing that's still off I can see is the span flush count from the lightstep sink.

asf-stripe commented:
I rebased this onto the bug fix for the write amplification and the `ssf.Samples` struct from #370, which takes care of the CPU usage and the lack of lightstep metrics (and, annoyingly, some other metrics that we didn't see drop off in our dashes)

asf-stripe commented:
Interesting results from testing veneur-proxy: It looks like the number of import times reported (`veneur_proxy.import.response_duration_ns.count`) went down by a factor of 5 - I don't yet know why that is. But in testing, I found some neat other result: There is a distinction in terms of GC happiness when it comes to different inputs: 
* spans via UDP pakcets GCs the most often (~0.8x/sec)
* framed spans via UNIX domain sockets (~0.4x/sec) and
* spans via the internal client almost don't GC at all (~0.1x/sec)

It all comes down to microsecond-level GC *times* though, so it's kinda moot (:

I'll investigate why the numbers are off now, though. Maybe something on master caused this.

asf-stripe commented:
So I found the reason why metrics on the proxy were off: This PR caused importing SSF packets with embedded metrics to race on map iteration/modification, which caused the local veneur to crash. I fixed that issue, but the counts on veneur-proxy are still off.

asf-stripe commented:
Updated to latest master, and added two bug fixes:

* The metrics extraction sink wouldn't actually count the number of spans/metrics.
* In parsing SSF metrics into internal ones, don't truncate the tags.

This results in the count of metrics proxied by host being tagged correctly again, and fixes the numbers reported by the metrics extraction sink. HOWEVER, it still looks like the veneur-proxy numbers are off by a pretty wide margin. This is weird, and I think to fully debug this, I'll have to add a metric to the client. Hrm.

asf-stripe commented:
I think the latest changes should fix all the problems I saw in testing this on veneur-proxies. That means that all the variants that we run veneur in are now reporting metrics correctly, as far as I can tell! (-:

I've also added a change to put span uniqueness tracking back, fixing up the double-count issue I saw when I ran this on veneur-proxies.

To that end: PTAL, @aditya-stripe?

asf-stripe commented:
Uh, ok, it looks like creating a HLL is racing ... with itself? And that's why tests are failing

```
15:00:48 WARNING: DATA RACE
15:00:48 Write at 0x000001763c58 by goroutine 367:
15:00:48   github.com/stripe/veneur/vendor/github.com/axiomhq/hyperloglog.new()
15:00:48       /go/src/github.com/stripe/veneur/vendor/github.com/axiomhq/hyperloglog/hyperloglog.go:50 +0x52
15:00:48   github.com/stripe/veneur/vendor/github.com/axiomhq/hyperloglog.New14()
15:00:48       /go/src/github.com/stripe/veneur/vendor/github.com/axiomhq/hyperloglog/hyperloglog.go:38 +0x33
15:00:48   github.com/stripe/veneur/vendor/github.com/axiomhq/hyperloglog.New()
15:00:48       /go/src/github.com/stripe/veneur/vendor/github.com/axiomhq/hyperloglog/hyperloglog.go:33 +0x2f
15:00:48   github.com/stripe/veneur/samplers.NewSet()
15:00:48       /go/src/github.com/stripe/veneur/samplers/samplers.go:273 +0x33
15:00:48   github.com/stripe/veneur.TestWorkerImportSet()
15:00:48       /go/src/github.com/stripe/veneur/worker_test.go:87 +0x229
15:00:48   testing.tRunner()
15:00:48       /usr/local/go/src/testing/testing.go:746 +0x16c
15:00:48 
15:00:48 Previous write at 0x000001763c58 by goroutine 351:
15:00:48   github.com/stripe/veneur/vendor/github.com/axiomhq/hyperloglog.new()
15:00:48       /go/src/github.com/stripe/veneur/vendor/github.com/axiomhq/hyperloglog/hyperloglog.go:50 +0x52
15:00:49   github.com/stripe/veneur/vendor/github.com/axiomhq/hyperloglog.New14()
15:00:49       /go/src/github.com/stripe/veneur/vendor/github.com/axiomhq/hyperloglog/hyperloglog.go:38 +0x33
15:00:49   github.com/stripe/veneur/vendor/github.com/axiomhq/hyperloglog.New()
15:00:49       /go/src/github.com/stripe/veneur/vendor/github.com/axiomhq/hyperloglog/hyperloglog.go:33 +0x2f
15:00:49   github.com/stripe/veneur/samplers.NewSet()
15:00:49       /go/src/github.com/stripe/veneur/samplers/samplers.go:273 +0x33
15:00:49   github.com/stripe/veneur.WorkerMetrics.Upsert()
15:00:49       /go/src/github.com/stripe/veneur/worker.go:122 +0x2d4
15:00:49   github.com/stripe/veneur.(*Worker).ProcessMetric()
15:00:49       /go/src/github.com/stripe/veneur/worker.go:192 +0x249
15:00:49   github.com/stripe/veneur.(*Worker).Work()
15:00:49       /go/src/github.com/stripe/veneur/worker.go:163 +0x44f
15:00:49   github.com/stripe/veneur.NewFromConfig.func1()
15:00:49       /go/src/github.com/stripe/veneur/server.go:233 +0x5e
```

asf-stripe commented:
(Rebuilding fixed this, but we should probably fix that race) (-:

ChimeraCoder commented:
This race failed twice now in the rebuild. I'm not sure why this code is triggering that particular data race more frequently, but we should definitely fix that before merging, to avoid making our builds flaky.

asf-stripe commented:
Tests pass on first try with #383 - seems like this ought to work now.

asf-stripe commented:
Rebased this onto latest master which should be free of race conditions and is building green - PTAL, @aditya-stripe! (:

ChimeraCoder commented on example_proxy.yaml:
```diff
@@ -6,8 +6,13 @@ http_address: "localhost:8127"
 
+# The address to which to send SSF spans and metrics - this is the
+# same format as on the veneur server's `ssf_listen_addresses`.
+ssf_destination_address: "udp://localhost:8200"
```
This should be `udp://localhost:8126`

asf-stripe commented:
I'd messed up the port example myself - copy/paste thinko from when it was just a statsd address. Thanks for the save, @krisreeves-stripe!! I'll self-approve and merge now (:
